1. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

2. Importing the data set


In [2]:
df = pd.read_csv("student_data.csv")

3. Displaying data - first 5 features


In [3]:
df.head()

Marital status  Application mode  Application order  Course  \
0              1                17                  5   171.0   
1              1                15                  1  9254.0   
2              1                 1                  5  9070.0   
3              1                17                  2  9773.0   
4              2                39                  1  8014.0   

   Daytime/evening attendance  Previous qualification  \
0                           1                       1   
1                           1                       1   
2                           1                       1   
3                           1                       1   
4                           0                       1   

   Previous qualification (grade)  Nacionality  Mother's qualification  \
0                           122.0            1                      19   
1                           160.0            1                       1   
2                           122.0            1                      37   
3                           122.0            1                      38   
4                           100.0            1                      37   

   Father's qualification  ...  Curricular units 2nd sem (credited)  \
0                    12.0  ...                                    0   
1                     3.0  ...                                    0   
2                    37.0  ...                                    0   
3                    37.0  ...                                    0   
4                    38.0  ...                                    0   

   Curricular units 2nd sem (enrolled)  \
0                                  0.0   
1                                  6.0   
2                                  6.0   
3                                  6.0   
4                                  6.0   

   Curricular units 2nd sem (evaluations) Curricular units 2nd sem (approved)  \
0                                       0                                   0   
1                                       6                                   6   
2                                       0                                   0   
3                                      10                                   5   
4                                       6                                   6   

   Curricular units 2nd sem (grade)  \
0                          0.000000   
1                         13.666667   
2                          0.000000   
3                         12.400000   
4                         13.000000   

   Curricular units 2nd sem (without evaluations) Unemployment rate  \
0                                               0              10.8   
1                                               0              13.9   
2                                               0              10.8   
3                                               0               9.4   
4                                               0              13.9   

   Inflation rate   GDP    Target  
0             1.4  1.74   Dropout  
1            -0.3  0.79  Graduate  
2             1.4  1.74   Dropout  
3            -0.8 -3.12  Graduate  
4            -0.3  0.79  Graduate  

[5 rows x 37 columns]

In [4]:
df.describe()

Application mode  Application order       Course  \
count       4424.000000        4424.000000  4423.000000   
mean          18.669078           1.727848  8856.458738   
std           17.484682           1.313793  2063.763479   
min            1.000000           0.000000    33.000000   
25%            1.000000           1.000000  9085.000000   
50%           17.000000           1.000000  9238.000000   
75%           39.000000           2.000000  9556.000000   
max           57.000000           9.000000  9991.000000   

       Daytime/evening attendance  Previous qualification  \
count                 4424.000000             4424.000000   
mean                     0.890823                4.577758   
std                      0.311897               10.216592   
min                      0.000000                1.000000   
25%                      1.000000                1.000000   
50%                      1.000000                1.000000   
75%                      1.000000                1.000000   
max                      1.000000               43.000000   

       Previous qualification (grade)  Nacionality  Mother's qualification  \
count                     4424.000000  4424.000000             4424.000000   
mean                       132.613314     1.873192               19.561935   
std                         13.188332     6.914514               15.603186   
min                         95.000000     1.000000                1.000000   
25%                        125.000000     1.000000                2.000000   
50%                        133.100000     1.000000               19.000000   
75%                        140.000000     1.000000               37.000000   
max                        190.000000   109.000000               44.000000   

       Father's qualification  Mother's occupation  ...  \
count             4423.000000          4424.000000  ...   
mean                22.271987            10.960895  ...   
std                 15.343244            26.418253  ...   
min                  1.000000             0.000000  ...   
25%                  3.000000             4.000000  ...   
50%                 19.000000             5.000000  ...   
75%                 37.000000             9.000000  ...   
max                 44.000000           194.000000  ...   

       Curricular units 1st sem (without evaluations)  \
count                                     4424.000000   
mean                                         0.137658   
std                                          0.690880   
min                                          0.000000   
25%                                          0.000000   
50%                                          0.000000   
75%                                          0.000000   
max                                         12.000000   

       Curricular units 2nd sem (credited)  \
count                          4424.000000   
mean                              0.541817   
std                               1.918546   
min                               0.000000   
25%                               0.000000   
50%                               0.000000   
75%                               0.000000   
max                              19.000000   

       Curricular units 2nd sem (enrolled)  \
count                          4423.000000   
mean                              6.231743   
std                               2.196038   
min                               0.000000   
25%                               5.000000   
50%                               6.000000   
75%                               7.000000   
max                              23.000000   

       Curricular units 2nd sem (evaluations)  \
count                             4424.000000   
mean                                 8.063291   
std                                  3.947951   
min                                  0.000000   
25%                                  6.000000   
50%                                  8.000000   
75%

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Marital status                                  4424 non-null   object 
 1   Application mode                                4424 non-null   int64  
 2   Application order                               4424 non-null   int64  
 3   Course                                          4423 non-null   float64
 4   Daytime/evening attendance                      4424 non-null   int64  
 5   Previous qualification                          4424 non-null   int64  
 6   Previous qualification (grade)                  4424 non-null   float64
 7   Nacionality                                     4424 non-null   int64  
 8   Mother's qualification                          4424 non-null   int64  
 9   Father's qualification                   

I will rename the curricular units features so they are easier to read in plain english, first I will change the first semester titles

In [6]:

df.rename(columns={'Curricular units 1st sem (credited)': 'Num of Modules 1st sem', 'Curricular units 1st sem (enrolled)': 'Enrolled Modules 1st sem','Curricular units 1st sem (evaluations)': 'Num of Assesments 1st sem','Curricular units 1st sem (approved)': 'Approved Modules 1st sem', 'Curricular units 1st sem (grade)': 'Module Grade 1st sem', 'Curricular units 1st sem (without evaluations)': 'Modules w/o Assesments 1st sem'}, inplace=True)



In [7]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                                          Non-Null Count  Dtype  
---  ------                                          --------------  -----  
 0   Marital status                                  4424 non-null   object 
 1   Application mode                                4424 non-null   int64  
 2   Application order                               4424 non-null   int64  
 3   Course                                          4423 non-null   float64
 4   Daytime/evening attendance                      4424 non-null   int64  
 5   Previous qualification                          4424 non-null   int64  
 6   Previous qualification (grade)                  4424 non-null   float64
 7   Nacionality                                     4424 non-null   int64  
 8   Mother's qualification                          4424 non-null   int64  
 9   Father's qualification                   

Now I'll rename the 2nd semester titles

In [8]:
df.rename(columns={'Curricular units 2nd sem (credited)': 'Num of Modules 2nd sem', 'Curricular units 2nd sem (enrolled)': 'Enrolled Modules 2nd sem','Curricular units 2nd sem (evaluations)': 'Num of Assesments 2nd sem','Curricular units 2nd sem (approved)': 'Approved Modules 2nd sem', 'Curricular units 2nd sem (grade)': 'Module Grade 2nd sem', 'Curricular units 2nd sem (without evaluations)': 'Modules w/o Assesments 2nd sem'}, inplace=True)




In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4424 entries, 0 to 4423
Data columns (total 37 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Marital status                  4424 non-null   object 
 1   Application mode                4424 non-null   int64  
 2   Application order               4424 non-null   int64  
 3   Course                          4423 non-null   float64
 4   Daytime/evening attendance      4424 non-null   int64  
 5   Previous qualification          4424 non-null   int64  
 6   Previous qualification (grade)  4424 non-null   float64
 7   Nacionality                     4424 non-null   int64  
 8   Mother's qualification          4424 non-null   int64  
 9   Father's qualification          4423 non-null   float64
 10  Mother's occupation             4424 non-null   int64  
 11  Father's occupation             4424 non-null   int64  
 12  Admission grade                 44

df.info allows us to see that there is 4424 observations and 37 features(columns)

It also shows the data type of each feature:
integers - fractional numbers
floats - whole numbers 
object - text 

In [10]:
df.isnull().sum()

Marital status                    0
Application mode                  0
Application order                 0
Course                            1
Daytime/evening attendance        0
Previous qualification            0
Previous qualification (grade)    0
Nacionality                       0
Mother's qualification            0
Father's qualification            1
Mother's occupation               0
Father's occupation               0
Admission grade                   1
Displaced                         0
Educational special needs         0
Debtor                            1
Tuition fees up to date           1
Gender                            0
Scholarship holder                0
Age at enrollment                 0
International                     1
Num of Modules 1st sem            0
Enrolled Modules 1st sem          0
Num of Assesments 1st sem         0
Approved Modules 1st sem          0
Module Grade 1st sem              1
Modules w/o Assesments 1st sem    0
Num of Modules 2nd sem      

is.null allows us to see any null values present in the data set, here we a few null values (1) present in some features 

We can use .unique() to show what all the null values are in the data set 

In [11]:
df['Course'].unique()

array([ 171., 9254., 9070., 9773., 8014., 9991., 9500., 9238., 9670.,
       9853., 9085., 9130., 9556., 9147., 9003.,   33., 9119.,   nan])

In [12]:
df["Father's qualification"].unique()

array([12.,  3., 37., 38.,  1., 19.,  5.,  4., 34.,  2., 39., 11.,  9.,
       nan, 36., 26., 40., 14., 20., 35., 41., 22., 13., 29., 43., 18.,
       42., 10.,  6., 30., 25., 44., 33., 27., 31.])

In [13]:
df["Admission grade"].unique()

array([127.3, 142.5, 124.8, 119.6, 141.5, 114.8, 128.4, 113.1, 129.3,
       123. , 130.6, 119.3, 130.2, 111.8, 137.1, 120.7, 137.4, 136.3,
       124.6, 120.3, 121.8, 125.5, 114.9, 123.9, 157. , 116.4, 131. ,
       122.1, 118.8, 150. , 130. , 138.8, 134.5, 131.4, 102.5, 128.8,
       122.9, 113.9, 120. , 121.1, 120.4, 100.6, 121.4, 109.7, 134.1,
       127.6, 132.4, 133.4, 126.1, 113.5, 121.3, 159.3, 129.1, 155.3,
       139.8, 115.2, 131.9, 126. , 120.9, 128.2, 120.1, 100. , 134. ,
       130.8, 135.8, 111.7, 132.9, 115.5, 106. , 117. , 110.2, 155.7,
       180.4, 110. , 161. , 117.6, 128.7, 112.2, 100.8, 105. , 114. ,
       137. , 124.9, 134.3, 111.5, 160. , 117.4, 122.2, 118.2, 106.7,
       108.2, 107. , 136.1, 115.3, 140.4, 113.4, 118.6, 122.3, 127.9,
       117.1, 145.3, 122.6, 128. , 123.7, 131.7, 133.2, 109.3, 113. ,
       157.9, 112.1, 174.7, 110.1,  99.7, 121. , 119.1, 124.7, 117.2,
       131.5, 121.7, 123.4, 132.8, 108.7, 138.1, 126.5, 127.4, 123.6,
       122. , 125.8,

In [14]:
df["Debtor"].unique()

array([ 0.,  1., nan])

In [15]:
df["Tuition fees up to date"].unique()

array(['1', '0', nan, '?'], dtype=object)

In [16]:
df["International"].unique()

array([ 0.,  1., nan])

In [18]:
df['Module Grade 1st sem'].unique()

array([ 0.        , 14.        , 13.42857143, 12.33333333, 11.85714286,
       13.3       , 13.875     , 11.4       , 13.21428571, 10.57142857,
       13.25      , 13.2       , 12.        , 13.30625   , 12.5       ,
       11.66666667, 11.4375    , 12.85714286, 13.375     , 13.29666667,
       11.6       , 11.375     , 12.66666667, 12.93333333, 12.83333333,
       11.33333333, 12.4       , 10.        , 11.        , 12.75      ,
       14.8       , 13.928     , 13.        , 11.5       , 13.51666667,
       13.66666667, 11.83333333, 14.71428571, 10.6       , 13.04444444,
       14.935     , 13.13333333, 11.9       , 12.57142857, 11.2       ,
       11.975     , 10.33333333, 10.5       , 14.84285714, 12.2       ,
       12.16666667, 13.08      , 10.66666667, 11.875     , 13.75      ,
       14.2       , 11.96      , 12.875     , 14.63157895, 13.26666667,
       13.85714286, 13.4       , 12.51666667, 11.61666667, 13.14285714,
       12.6       , 10.8       , 12.8       , 14.16666667, 12.1 

In [19]:
df['Enrolled Modules 2nd sem'].unique()

array([ 0.,  6.,  5.,  8.,  7., 11., 12.,  9., 13., 19.,  3., 10.,  4.,
       nan, 17.,  2.,  1., 14., 15., 16., 23., 18., 21.])

In [21]:
df['Module Grade 2nd sem'].unique() 

array([ 0.        , 13.66666667, 12.4       , 13.        , 11.5       ,
       14.345     , 14.14285714, 13.5       , 14.2       , 13.21428571,
       11.        , 12.        , 14.545     , 12.25      , 11.425     ,
       12.85714286, 12.28571429, 14.11428571, 13.28571429, 12.33333333,
       13.71666667, 10.57142857, 13.4       , 14.375     , 13.42857143,
       10.        , 11.6       , 15.8       , 14.34285714, 13.8       ,
       14.33333333, 13.2       , 13.77142857, 12.83333333, 14.16666667,
       11.83333333, 13.595     , 12.9       , 14.935     , 12.22142857,
       12.42857143, 10.75      , 10.8       , 13.75      , 10.85714286,
       13.33333333, 15.05555556, 11.66666667, 12.2       , 13.3       ,
       11.33333333, 13.08      , 11.2       , 13.25      , 14.47777778,
       14.23076923, 12.27272727, 12.3       , 15.        , 12.8       ,
       12.57142857, 12.51666667, 10.66666667, 11.4       , 14.        ,
       11.25      , 13.77777778, 15.01111111, 13.74285714, 14.6 

In [22]:
missing_value=['n.a.','?','NA','n/a','na','--','nan']

In [25]:
newdf=pd.read_csv('student_data.csv', na_values=missing_value)

In [26]:
newdf.isnull().sum()

Marital status                                    1
Application mode                                  0
Application order                                 0
Course                                            1
Daytime/evening attendance                        0
Previous qualification                            0
Previous qualification (grade)                    0
Nacionality                                       0
Mother's qualification                            0
Father's qualification                            1
Mother's occupation                               0
Father's occupation                               0
Admission grade                                   1
Displaced                                         1
Educational special needs                         0
Debtor                                            1
Tuition fees up to date                           2
Gender                                            0
Scholarship holder                                0
Age at enrol